napari/_qt/qt_main_window.py

```diff
@@ -440,6 +452,131 @@ def __init__(self, viewer, *, show: bool = True):
         if show:
             self.show()
 
+    def _add_plugins_theme(self, event=None):
+        """Update stylesheet/icon/theme."""
+        from ..utils.theme import register_theme
+
+        # clear out all data to make sure it's up to date (e.g. after
+        # plugin was disabled or unregistered)
+        plugin_manager._theme_data.clear()
+        plugin_manager.discover_themes()
+        for theme_name, theme_data in plugin_manager.iter_themes():
+            register_theme(theme_name, theme_data)
+        self._rebuild_theme_settings()
+
+    def _remove_plugins_theme(self, event=None):
+        """Remove plugin."""
+        from ..utils.theme import unregister_theme
+
+        plugin_name = event.value
+
+        # since its possible that disabled/removed plugin was providing the
+        # current theme, we check for this explicitly and if this the case,
+        # theme is automatically changed to default `dark` theme
+        settings = get_settings()
+        current_theme = settings.appearance.theme
+        if (
+            settings.appearance.theme
+            in plugin_manager._theme_data[plugin_name]
+        ):
+            self.qt_viewer.viewer.theme = "dark"
+            warnings.warn(
+                message=trans._(
+                    "The current theme {current_theme!r} was provided by the"
+                    " plugin {plugin_name!r} which was disabled or removed."
+                    " Switched theme to the default",
+                    deferred=True,
+                    plugin_name=plugin_name,
+                    current_theme=current_theme,
+                )
+            )
+
+        # unregister all themes that were provided by the plugins
+        for theme_name in plugin_manager._theme_data[plugin_name].keys():
+            unregister_theme(theme_name)
+
+        self._rebuild_theme_settings()
+        self._update_theme()
+
+    def _setup_existing_themes(self):
+        """This function is only executed once at the startup of napari
+        to connect events to themes that have not been connected yet."""
+        for theme in _themes.values():
+            self._connect_theme(theme)
+
+    def _add_theme(self, event):
+        theme = event.value
+        self._connect_theme(theme)
+
+    def _connect_theme(self, theme):
+        # connect events to update theme. Here, we don't want to pass the event
+        # since it won't have the right `value` attribute.
+        theme.events.background.connect(lambda _: self._update_theme())
```

Comment:
fyi, I believe you can connect directly to `theme.events.connect` so :`theme.events.connect(lambda _: self._update_theme())` gets them all done in one go  (or at least I think 🤔)

Reply:
That's true. I did think of doing it like this and we still can. I've decided against it because the `icon` attribute is a special case where the `_update_theme` method would do nothing.